File: javascript/pig-latin/pig-latin.js
const pigLatin = function() {};

pigLatin.translate = function(str) {
  const words    = str.split(' ');
  let translated = '';
  const vowels   = /[aeiou]/gi;
  const result   = words.map( word => {
    const firstConsonant = word.indexOf(word.match(vowels)[0]);

    function vowel(word) {
      translated = word + 'ay';
      return translated;
    }

    function consonant(word) {
      // check is q is preceding a vowel (queen and sqaure example)
      if ( word[firstConsonant-1] === 'q' ) return word.substr(firstConsonant+1) + word.substr(0, firstConsonant+1) + 'ay';
      translated = word.substr(firstConsonant) + word.substr(0, firstConsonant) + 'ay';
      return translated;
    }

    return firstConsonant ? consonant(word) : vowel(word);
  });
  return result.join(' ');
};

module.exports = pigLatin;

File: ruby/pig-latin/pig_latin.rb

class PigLatin

  VOWELS = ['a', 'e', 'i', 'o', 'u', 'y']
  def self.translate sentence
    words = sentence.split(" ")
    words.map! do |word|
      result = word.split("")
      if VOWELS.index(result[0])
        result.join + "ay"
      elsif word.index("qu") == 0
        result.rotate!(2)
        result.join + "ay"
      else
        result.rotate!(1)
        if result.index("q") == 0 && result.index("u") == 1
          result.rotate!(2)
          result.join + "ay"
        end
        while VOWELS.index(result[0]) == nil
          result.rotate!(1)
        end
        result.join + "ay"
      end # if
    end # map
    words.join(" ")
  end # def

end # class
